File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/controls/CustomPercent.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.controls;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;

import java.util.regex.Matcher;
import java.util.regex.Pattern;


/**
 * Implements custom view to graph percentage around text.
 * @author Eric Ruck
 */
public class CustomPercent extends android.support.v7.widget.AppCompatTextView {

	private static final float STROKE_WIDTH = 4;
	private static final int FULL_COLOR = Color.LTGRAY;
	private static final int PCT_COLOR = Color.BLUE;

	private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);


	public CustomPercent(Context context, AttributeSet attrs) {
		super(context, attrs);
		paint.setStyle(Paint.Style.STROKE);
		paint.setStrokeWidth(STROKE_WIDTH);
	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);

		// Is there a percentage in our text?
		Pattern pattern = Pattern.compile("^\\d+");
		Matcher matcher = pattern.matcher(getText());
		if (matcher.find()) {
			int percent = Integer.parseInt(matcher.group());
			if ((percent >= 0) && (percent <= 100)) {
				// Draw the full ring
				float half = STROKE_WIDTH / 2;
				float width = getWidth() - STROKE_WIDTH;
				float height = getHeight() - STROKE_WIDTH;
				paint.setColor(percent == 100 ? PCT_COLOR : FULL_COLOR);
				canvas.drawOval(half, half, width, height, paint);
				if ((percent > 0) && (percent < 100)) {
					// Draw percentage arc
					float sweep= 360 * percent / 100;
					paint.setColor(PCT_COLOR);
					canvas.drawArc(half, half, width, height, 270, sweep, false, paint);
				}
			}
		}
	}
}

File: ios/Mobile Client/Mobile Client/controls/CustomPercent.swift
//
//  CustomPercent.swift
//  Mobile Client
//
//  Created by Eric Ruck on 8/13/18.
//  Copyright 2018-2019 AuditPro. All rights reserved.
//

import UIKit


/**
 * Draws the percentage around the outside of the control.
 * @author Eric Ruck
 */
class CustomPercent: UILabel {
	private static let STROKE_WIDTH: CGFloat = 2.0
	private static let FULL_COLOR = UIColor.lightGray
	private static let PCT_COLOR = UIColor.blue

	/**
	 * Custom painting for this label control.
	 * @param rect Drawing rectangle
	 */
    override func draw(_ rect: CGRect) {
        // Default first
        super.draw(rect)
		guard let text = self.text else {
			// Unable to parse percentage
			return
		}
		if	let range = text.range(of: "^\\d+", options: .regularExpression, range: nil, locale: nil),
		 	let percent = Int(text[range]) {
			// Draw the percentage in our text
	 		let radius = (bounds.size.width - CustomPercent.STROKE_WIDTH) / 2
	 		let offset = radius + (CustomPercent.STROKE_WIDTH / 2)
	 		let center = CGPoint(x: offset, y: offset)
		 	if (percent >= 0) && (percent < 100) {
		 		// Draw full arc
		 		let path = UIBezierPath(arcCenter: center, radius: radius, startAngle: 0, endAngle: .pi * 2, clockwise: true)
		 		path.lineWidth = CustomPercent.STROKE_WIDTH
		 		CustomPercent.FULL_COLOR.setStroke()
		 		path.stroke()
			}
		 	if (percent > 0) && (percent <= 100) {
		 		// Draw percent arc
		 		let startAngle = CGFloat(.pi * 1.5)
		 		let endAngle = CGFloat(percent - 25) / 50.0 * .pi
		 		let path = UIBezierPath(arcCenter: center, radius: radius, startAngle: startAngle, endAngle: endAngle, clockwise: true)
		 		path.lineWidth = CustomPercent.STROKE_WIDTH
		 		CustomPercent.PCT_COLOR.setStroke()
		 		path.stroke()
			}
		}
    }
}
